ultraplot/ticker.py

```diff
@@ -822,6 +846,492 @@ def __call__(self, x, pos=None):  # noqa: U100
+            The calendar for interpreting numeric tick values.
+        units : str, default: 'days since 2000-01-01'
+            The time units for interpreting numeric tick values.
+        """
+        if cftime is None:
+            raise ModuleNotFoundError("cftime is required for CFDatetimeFormatter.")
+        self._format = fmt
+        self._calendar = calendar
+        self._units = units
+
+    def __call__(self, x, pos=None):
+        if isinstance(x, cftime.datetime):
+            dt = x
+        else:
+            dt = cftime.num2date(x, self._units, calendar=self._calendar)
+        return dt.strftime(self._format)
+
+
+class AutoCFDatetimeFormatter(mticker.Formatter):
+    """Automatic formatter for `cftime.datetime` data."""
+
+    def __init__(self, locator, calendar, time_units=None):
+        self.locator = locator
+        self.calendar = calendar
+        self.time_units = time_units or _TIME_UNITS
+
+    def pick_format(self, resolution):
+        return _RESOLUTION_TO_FORMAT[resolution]
+
+    def __call__(self, x, pos=0):
+        format_string = self.pick_format(self.locator.resolution)
+        if isinstance(x, cftime.datetime):
+            dt = x
+        else:
+            dt = cftime.num2date(x, self.time_units, calendar=self.calendar)
+        return dt.strftime(format_string)
+
+
+class AutoCFDatetimeLocator(mticker.Locator):
+    """Determines tick locations when plotting `cftime.datetime` data."""
+
+    real_world_calendars = (
+        "gregorian",
+        "julian",
+        "proleptic_gregorian",
+        "standard",
+    )
+
+    def __init__(self, maxticks=None, calendar="standard", date_unit=None, minticks=3):
+        super().__init__()
+        self.minticks = minticks
+        # These are thresholds for *determining resolution*, NOT directly for MaxNLocator tick count
+        self.maxticks = {
+            "YEARLY": 1,
+            "MONTHLY": 12,
+            "DAILY": 8,
+            "HOURLY": 11,
+            "MINUTELY": 1,
+            "SECONDLY": 11,  # Added for completeness, though not a resolution threshold
+        }
+        if maxticks is not None:
+            if isinstance(maxticks, dict):
+                self.maxticks.update(maxticks)
+            else:
+                # If a single value is provided for maxticks, apply it to all resolution *thresholds*
+                self.maxticks = dict.fromkeys(self.maxticks.keys(), maxticks)
+
+        self.calendar = calendar
+        self.date_unit = date_unit or _TIME_UNITS
+        if not self.date_unit.lower().startswith("days since"):
+            emsg = "The date unit must be days since for a NetCDF time locator."
+            raise ValueError(emsg)
+        self.resolution = _DEFAULT_RESOLUTION
+        self._cached_resolution = {}
+
+        # New: Default max ticks for Matplotlib's MaxNLocator, independent of resolution thresholds
+        self._max_display_ticks = 7  # A reasonable number for actual axis ticks
```

Comment:
Does this need to be in the config file?

Reply:
Sure let's do it.